init.lua

```diff
@@ -30,6 +30,20 @@ local function destroy(handle)
 end
 ffi.gc(cudnn.handle, destroy)
 
+cudnn.typemap = {
+   ['torch.CudaHalfTensor']   = 'CUDNN_DATA_HALF',
+   ['torch.CudaTensor']       = 'CUDNN_DATA_FLOAT',
+   ['torch.CudaDoubleTensor'] = 'CUDNN_DATA_DOUBLE',
+}
+
+-- TODO: determine if device supports true half and use true half on it
```

Comment:
OK I had to increase epsilon for half from 3.0 to 4.0 and now all tests are passing on Maxwell.
On Pascal, however, all convolutions are failing - our CUDNN experts say only some combinations of sizes are supported for fp16 convolution descriptors, so we use fixed sizes instead of radon ones in the test (still waiting on the list of supported sizes):

boris@snikolaevX99A:~/git/distro/extra/cudnn$ th ./test/test.lua 
Running test on device: 1
Testing torch.CudaHalfTensor
Running 38 tests
 1/38 VolumetricMaxPooling_batch ........................................ [PASS]
 2/38 SpatialMaxPooling_single .......................................... [PASS]
 3/38 SpatialCrossEntropyCriterion ...................................... [PASS]
 4/38 SpatialMaxPooling_batch ........................................... [PASS]
 5/38 SpatialAveragePooling_batch ....................................... [PASS]
 6/38 LogSoftMax_single ................................................. [PASS]
 7/38 functional_bias2D ................................................. [PASS]
 8/38 SpatialBatchNormalization ......................................... [PASS]
 9/38 SpatialConvolution_forward_single ................................. [ERROR]
10/38 functional_convolution2d .......................................... [ERROR]
11/38 SpatialFullConvolution_backward_batch ............................. [ERROR]
12/38 TemporalConvolution_padding_batch ................................. [ERROR]
13/38 SpatialConvolution_backward_single ................................ [ERROR]
14/38 Tanh_single ....................................................... [PASS]
15/38 ClippedReLU_single ................................................ [PASS]
16/38 ReLU_single ....................................................... [PASS]
17/38 Sigmoid_single .................................................... [PASS]
18/38 VolumetricConvolution_forward_single .............................. [ERROR]
19/38 Tanh_batch ........................................................ [PASS]
20/38 SoftMax_batch ..................................................... [PASS]
21/38 SpatialCrossMapLRN_batch .......................................... [PASS]
22/38 VolumetricMaxPooling_single ....................................... [FAIL]
23/38 VolumetricBatchNormalization ...................................... [PASS]
24/38 ReLU_batch ........................................................ [PASS]
25/38 SpatialConvolution_backward_batch ................................. [ERROR]
26/38 TemporalConvolution_batch ......................................... [ERROR]
27/38 SpatialFullConvolution_forward_batch .............................. [ERROR]
28/38 Sigmoid_batch ..................................................... [PASS]
29/38 SpatialLogSoftMax ................................................. [PASS]
30/38 TemporalConvolution_reduceBatchSize ............................... [ERROR]
31/38 LogSoftMax_batch .................................................. [PASS]
32/38 BatchNormalization ................................................ [PASS]
33/38 TemporalConvolution_single ........................................ [ERROR]
34/38 functional_maxpooling2d ........................................... [PASS]
35/38 SpatialConvolution_forward_batch .................................. [ERROR]
36/38 SpatialAveragePooling_single ...................................... [PASS]
37/38 VolumetricConvolution_backward_single ............................. [ERROR]
38/38 SoftMax_single .................................................... [PASS]
Completed 173 asserts in 38 tests with 1 failure and 13 errors
--------------------------------------------------------------------------------
SpatialConvolution_forward_single
 Function call failed
/usr/share/lua/5.1/cudnn/init.lua:88: Error in CuDNN: CUDNN_STATUS_BAD_PARAM (cudnnGetConvolutionBackwardFilterAlgorithm)
stack traceback:
        [C]: in function 'error'
        /usr/share/lua/5.1/cudnn/init.lua:88: in function 'errcheck'
        /usr/share/lua/5.1/cudnn/SpatialConvolution.lua:253: in function 'createIODescriptors'
        /usr/share/lua/5.1/cudnn/SpatialConvolution.lua:364: in function 'forward'
        ./test/test.lua:176: in function 'test'
        ./test/test.lua:189: in function <./test/test.lua:155>
        [C]: in function 'xpcall'
        /usr/share/lua/5.1/torch/Tester.lua:477: in function '_pcall'
        /usr/share/lua/5.1/torch/Tester.lua:436: in function '_run'
        /usr/share/lua/5.1/torch/Tester.lua:355: in function 'run'
        ./test/test.lua:1522: in main chunk
        [C]: in function 'dofile'
        /usr/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
        [C]: at 0x00406670

--------------------------------------------------------------------------------
functional_convolution2d
 Function call failed
/usr/share/lua/5.1/cudnn/init.lua:88: Error in CuDNN: CUDNN_STATUS_INVALID_VALUE (cudnnConvolutionForward)
stack traceback:
        [C]: in function 'error'
        /usr/share/lua/5.1/cudnn/init.lua:88: in function 'errcheck'
        /usr/share/lua/5.1/cudnn/functional.lua:117: in function 'Convolution2D_updateOutput'
        ./test/test.lua:1464: in function <./test/test.lua:1454>
        [C]: in function 'xpcall'
        /usr/share/lua/5.1/torch/Tester.lua:477: in function '_pcall'
        /usr/share/lua/5.1/torch/Tester.lua:436: in function '_run'
        /usr/share/lua/5.1/torch/Tester.lua:355: in function 'run'
        ./test/test.lua:1522: in main chunk